test/statusline.cpp

```diff
@@ -0,0 +1,202 @@
+
+			THEN("the message is passed to the IStatus implementation") {
+				REQUIRE(status_message_handler.last_message == "this is a status");
+				REQUIRE_FALSE(status_message_handler.last_string_was_error);
+			}
+		}
+
+		WHEN("show_error() is called") {
+			status_line.show_error("this is an error");
+
+			THEN("the message is passed to the IStatus implementation") {
+				REQUIRE(status_message_handler.last_message == "this is an error");
+				REQUIRE(status_message_handler.last_string_was_error);
+			}
+		}
+
+		WHEN("show_message_until_finished() is called") {
+			auto message_lifetime = status_line.show_message_until_finished("test message...");
+
+			THEN("the message is passed to the IStatus implementation") {
+				REQUIRE(status_message_handler.last_message == "test message...");
+				REQUIRE_FALSE(status_message_handler.last_string_was_error);
+			}
+
+			WHEN("the message_lifetime object goes out of scope") {
+				message_lifetime.reset();
+
+				THEN("the message is cleared") {
+					REQUIRE(status_message_handler.last_message == "");
+					REQUIRE_FALSE(status_message_handler.last_string_was_error);
+				}
+			}
+		}
+	}
+
+	GIVEN("a previous message or error") {
+		status_line.show_message("regular status");
+		status_line.show_error("error message");
+
+		WHEN("show_message() is called") {
+			status_line.show_message("this is a status");
+
+			THEN("the message is passed to the IStatus implementation") {
+				REQUIRE(status_message_handler.last_message == "this is a status");
+				REQUIRE_FALSE(status_message_handler.last_string_was_error);
+			}
+		}
+
+		WHEN("show_error() is called") {
+			status_line.show_error("this is an error");
+
+			THEN("the message is passed to the IStatus implementation") {
+				REQUIRE(status_message_handler.last_message == "this is an error");
+				REQUIRE(status_message_handler.last_string_was_error);
+			}
+		}
+
+		WHEN("show_message_until_finished() is called") {
+			auto message_lifetime = status_line.show_message_until_finished("test message...");
+
+			THEN("the message is passed to the IStatus implementation") {
+				REQUIRE(status_message_handler.last_message == "test message...");
+				REQUIRE_FALSE(status_message_handler.last_string_was_error);
+			}
+
+			WHEN("the message_lifetime object goes out of scope") {
+				message_lifetime.reset();
+
+				THEN("the message is cleared") {
+					REQUIRE(status_message_handler.last_message == "");
+					REQUIRE_FALSE(status_message_handler.last_string_was_error);
+				}
+			}
+		}
+	}
+
+	GIVEN("a previous show_message_until_finished call") {
+		const auto previous_message_lifetime =
+			status_line.show_message_until_finished("previous message...");
+
+		WHEN("show_message() is called") {
+			status_line.show_message("this is a status");
+
+			THEN("the message is passed to the IStatus implementation") {
+				REQUIRE(status_message_handler.last_message == "this is a status");
+				REQUIRE_FALSE(status_message_handler.last_string_was_error);
+			}
```

Comment:
I think this should also check what happens if `previous_message_lifetime` ends after a call to `show_message()`.